mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala

```diff
@@ -0,0 +1,158 @@
+
+  /** @group getParam */
+  def getUpperInclusive: Boolean = $(upperInclusive)
+
+  /** @group setParam */
+  def setUpperInclusive(value: Boolean): this.type = set(upperInclusive, value)
+
+  /** @group setParam */
+  def setInputCol(value: String): this.type = set(inputCol, value)
+
+  /** @group setParam */
+  def setOutputCol(value: String): this.type = set(outputCol, value)
+
+  override def transform(dataset: DataFrame): DataFrame = {
+    transformSchema(dataset.schema)
+    val wrappedSplits = Array(Double.MinValue) ++ $(splits) ++ Array(Double.MaxValue)
+    val bucketizer = udf { feature: Double =>
+      Bucketizer
+        .binarySearchForBuckets(wrappedSplits, feature, $(lowerInclusive), $(upperInclusive)) }
+    val newCol = bucketizer(dataset($(inputCol)))
+    val newField = prepOutputField(dataset.schema)
+    dataset.withColumn($(outputCol), newCol.as($(outputCol), newField.metadata))
+  }
+
+  private def prepOutputField(schema: StructType): StructField = {
+    val attr = new NominalAttribute(
+      name = Some($(outputCol)),
+      isOrdinal = Some(true),
+      values = Some($(splits).map(_.toString)))
+
+    attr.toStructField()
+  }
+
+  override def transformSchema(schema: StructType): StructType = {
+    SchemaUtils.checkColumnType(schema, $(inputCol), DoubleType)
+    require(schema.fields.forall(_.name != $(outputCol)),
+      s"Output column ${$(outputCol)} already exists.")
+    StructType(schema.fields :+ prepOutputField(schema))
+  }
+}
+
+object Bucketizer {
+  /**
+   * The given splits should match 1) its size is larger than zero; 2) it is ordered in a strictly
+   * increasing way.
+   */
+  private def checkSplits(splits: Array[Double]): Boolean = {
+    if (splits.size == 0) false
+    else if (splits.size == 1) true
+    else {
+      splits.foldLeft((true, Double.MinValue)) { case ((validator, prevValue), currValue) =>
+        if (validator && prevValue < currValue) {
+          (true, currValue)
+        } else {
+          (false, currValue)
+        }
+      }._1
+    }
+  }
+
+  /**
+   * Binary searching in several buckets to place each data point.
+   */
+  private[feature] def binarySearchForBuckets(
+      splits: Array[Double],
+      feature: Double,
+      lowerInclusive: Boolean,
+      upperInclusive: Boolean): Double = {
+    if ((feature < splits.head && !lowerInclusive) || (feature > splits.last && !upperInclusive)) {
+      throw new Exception(s"Feature $feature out of bound, check your features or loose the" +
+        s" lower/upper bound constraint.")
+    }
+    var left = 0
+    var right = splits.length - 2
+    while (left <= right) {
+      val mid = left + (right - left) / 2
+      val split = splits(mid)
+      if ((feature >= split) && (feature < splits(mid + 1))) {
+        return mid
+      } else if (feature < split) {
+        right = mid - 1
+      } else {
+        left = mid + 1
+      }
+    }
+    throw new Exception(s"Failed to find a bucket for feature $feature.")
```

Comment:
This should probably be a RuntimeException.  Can it also please say something like "unexpected error?"